src/Processors/Formats/Impl/CSVRowInputFormat.cpp

```diff
@@ -155,15 +165,17 @@ void CSVRowInputFormat::readPrefix()
     size_t num_columns = data_types.size();
     const auto & header = getPort().getHeader();
 
-    if (with_names)
+    /// This is a bit of abstraction leakage, but we have almost the same code in other places.
+    /// Thus, we check if this InputFormat is working with the "real" beginning of the data in case of parallel parsing.
+    if (with_names && getCurrentUnitNumber() == 0)
```

Comment:
Probably you also need an  assertion somewhere that `getCurrentUnitNumber() == 0 || column_mapping->is_set`.

This inverted control flow looks somewhat bad, did you consider having all the necessary logic in the ParallelFormatParser instead? I.e. you can add special handling of the first processing unit there -- wait for it and extract columns from it, and spread it to all other units. This is basically the same thing you're doing, but it won't be mixed across layers.

Reply:
All necessary logic is already implemented in ParallelParsingInputFormat, e.g. waiting for first piece of data to be processed, extracting ColumnMapping and propagating it to other parsers. 

This `hack` is needed to not to read "names header" in other parsers, because there are no "names header"